h2/src/main/org/h2/index/IndexCursor.java

```diff
@@ -210,22 +210,29 @@ private SearchRow getSpatialSearchRow(SearchRow row, int columnId, Value v) {
             v = ((ValueGeometry) v.convertTo(Value.GEOMETRY)).
                     getEnvelopeUnion(vg);
         }
-        if (columnId < 0) {
+        if (columnId == SearchRow.ROWID_INDEX) {
             row.setKey(v.getLong());
         } else {
             row.setValue(columnId, v);
         }
         return row;
     }
 
-    private SearchRow getSearchRow(SearchRow row, int columnId, Value v,
-            boolean max) {
+    private SearchRow getSearchRow(SearchRow row, int columnId, Value v, boolean max) {
+        Column column = columnId == SearchRow.ROWID_INDEX ?
+                                table.getRowIdColumn() :
+                                table.getColumn(columnId);
+        int vType = v.getType();
+        int resType = Value.getHigherOrder(column.getType(), vType);
+        if(vType != resType) {
+            v = column.convert(v, session.getDatabase().getMode());
+        }
```

Comment:
this new chunk of logic seems like it need some explanation beyond "replace magic constant" ? is it fixing a bug?

Reply:
It is definitely unrelated to "magic value", but it was long ago, and unfortunately, I do not have a test case for it, but yes, that was a bug, and as far as I can remember, it was about type mismatch between index column and provided value, so comparison in getMax() misbehaved. It just do value type conversion to column's type.

Reply:
I am retracting this now, since I really can't find the case (i think it was ENUM-related)